File: Display.py
import math
from PIL import Image, ImageDraw, ImageFont


class Display:
	_font_cache = {}

	def __init__(self, size=240):
		self.WIDTH = size
		self.HEIGHT = size
		self.buffer = bytearray(self.WIDTH * self.HEIGHT * 4)
		self.clear()

	def clear(self):
		for i in range(0, len(self.buffer), 4):
			self.buffer[i]	   = 0
			self.buffer[i + 1] = 0
			self.buffer[i + 2] = 0
			self.buffer[i + 3] = 0

	def fill(self, color=0x000000):
		r = (color >> 16) & 0xFF
		g = (color >> 8) & 0xFF
		b = color & 0xFF
		for i in range(0, len(self.buffer), 4):
			self.buffer[i]	   = r
			self.buffer[i + 1] = g
			self.buffer[i + 2] = b
			self.buffer[i + 3] = 255

	def _blend(self, idx, fr, fg, fb, alpha):
		if alpha <= 0.0:
			return
		if alpha > 1.0:
			alpha = 1.0

		ba = self.buffer[idx + 3] / 255.0
		outA = alpha + ba * (1.0 - alpha)

		if outA <= 0.0:
			self.buffer[idx]	 = 0
			self.buffer[idx + 1] = 0
			self.buffer[idx + 2] = 0
			self.buffer[idx + 3] = 0
			return

		if alpha >= 1.0:
			self.buffer[idx]	 = fr
			self.buffer[idx + 1] = fg
			self.buffer[idx + 2] = fb
			self.buffer[idx + 3] = 255
			return

		br = self.buffer[idx]
		bg = self.buffer[idx + 1]
		bb = self.buffer[idx + 2]
		remainder = ba * (1.0 - alpha)

		outR = (fr * alpha + br * remainder) / outA
		outG = (fg * alpha + bg * remainder) / outA
		outB = (fb * alpha + bb * remainder) / outA

		self.buffer[idx]	 = int(outR + 0.5)
		self.buffer[idx + 1] = int(outG + 0.5)
		self.buffer[idx + 2] = int(outB + 0.5)
		self.buffer[idx + 3] = int(outA * 255.0 + 0.5)

	@staticmethod
	def _pointToSegmentDistance(px, py, x1, y1, x2, y2):
		dx = x2 - x1
		dy = y2 - y1
		lenSq = dx * dx + dy * dy
		if lenSq == 0.0:
			ex = px - x1
			ey = py - y1
			return math.sqrt(ex * ex + ey * ey)
		t = ((px - x1) * dx + (py - y1) * dy) / lenSq
		if t < 0.0:
			t = 0.0
		if t > 1.0:
			t = 1.0
		projx = x1 + t * dx
		projy = y1 + t * dy
		ex = px - projx
		ey = py - projy
		return math.sqrt(ex * ex + ey * ey)

	@staticmethod
	def _pointToArcDistance(px, py, cx, cy, radius, startAngle, endAngle, halfWidth):
		if endAngle - startAngle >= 360.0:
			dx = px - cx
			dy = py - cy
			distToCenter = math.sqrt(dx*dx + dy*dy)
			return abs(distToCenter - radius)

		dx = px - cx
		dy = py - cy
		distToCenter = math.sqrt(dx * dx + dy * dy)

		sr = math.radians(startAngle)
		er = math.radians(endAngle)
		sr = sr % (2.0 * math.pi)
		er = er % (2.0 * math.pi)

		if distToCenter == 0.0:
			a0 = 0.0
			if startAngle <= endAngle:
				inRange = (a0 >= startAngle and a0 <= endAngle)
			else:
				inRange = (a0 >= startAngle or a0 <= endAngle)
		else:
			angle = math.atan2(dy, dx)
			if angle < 0.0:
				angle += 2.0 * math.pi

			if sr <= er:
				inRange = (angle >= sr and angle <= er)
			else:
				inRange = (angle >= sr or angle <= er)

		if inRange:
			return abs(distToCenter - radius)

		innerR = radius - halfWidth
		outerR = radius + halfWidth

		sxi = cx + innerR * math.cos(sr)
		syi = cy + innerR * math.sin(sr)
		sxo = cx + outerR * math.cos(sr)
		syo = cy + outerR * math.sin(sr)

		exi = cx + innerR * math.cos(er)
		eyi = cy + innerR * math.sin(er)
		exo = cx + outerR * math.cos(er)
		eyo = cy + outerR * math.sin(er)

		d1 = Display._pointToSegmentDistance(px, py, sxi, syi, sxo, syo)
		d2 = Display._pointToSegmentDistance(px, py, exi, eyi, exo, eyo)
		dmin = d1 if d1 < d2 else d2

		return dmin + halfWidth

	def drawLine(self, x1, y1, x2, y2, thickness, color):
		fr = (color >> 16) & 0xFF
		fg = (color >> 8) & 0xFF
		fb = color & 0xFF

		hw = thickness * 0.5
		xmin = min(x1, x2) - int(hw + 1.0)
		xmax = max(x1, x2) + int(hw + 1.0)
		ymin = min(y1, y2) - int(hw + 1.0)
		ymax = max(y1, y2) + int(hw + 1.0)

		if xmin < 0: xmin = 0
		if ymin < 0: ymin = 0
		if xmax >= self.WIDTH:  xmax = self.WIDTH - 1
		if ymax >= self.HEIGHT: ymax = self.HEIGHT - 1

		for y in range(ymin, ymax + 1):
			for x in range(xmin, xmax + 1):
				px = x + 0.5
				py = y + 0.5
				dist = self._pointToSegmentDistance(px, py,
													x1 + 0.5, y1 + 0.5,
													x2 + 0.5, y2 + 0.5)
				r = dist - hw
				if r <= -1.0:
					alpha = 1.0
				elif r >= 1.0:
					alpha = 0.0
				else:
					alpha = 1.0 - (r + 1.0) * 0.5

				if alpha <= 0.0:
					continue

				idx = (y * self.WIDTH + x) * 4
				self._blend(idx, fr, fg, fb, alpha)

	def drawArc(self, cx, cy, radius, startAngle, endAngle, thickness, color, stripe=False, stripeColor=0xffffff):
		fr = (color >> 16) & 0xFF
		fg = (color >> 8) & 0xFF
		fb = color & 0xFF
		hw = thickness * 0.5
		margin = int(hw + 1.0)
		xmin = int(cx - radius - margin)
		xmax = int(cx + radius + margin)
		ymin = int(cy - radius - margin)
		ymax = int(cy + radius + margin)
		if xmin < 0: xmin = 0
		if ymin < 0: ymin = 0
		if xmax >= self.WIDTH:  xmax = self.WIDTH - 1
		if ymax >= self.HEIGHT: ymax = self.HEIGHT - 1
		for y in range(ymin, ymax + 1):
			for x in range(xmin, xmax + 1):
				px = x + 0.5
				py = y + 0.5
				dist = self._pointToArcDistance(px, py, cx + 0.5, cy + 0.5, float(radius), startAngle, endAngle, hw)
				r = dist - hw
				if r <= -1.0:
					alpha = 1.0
				elif r >= 1.0:
					alpha = 0.0
				else:
					alpha = 1.0 - (r + 1.0) * 0.5
				if alpha <= 0.0:
					continue
				idx = (y * self.WIDTH + x) * 4
				self._blend(idx, fr, fg, fb, alpha)

		if stripe:
			halfWidth = thickness / 2.0
			innerR = radius - halfWidth
			outerR = radius + halfWidth
			stripe_spacing_deg = 6.0
			gridOrigin = -90.0
			n0 = math.ceil((startAngle - gridOrigin) / stripe_spacing_deg)
			a = gridOrigin + n0 * stripe_spacing_deg
			while a <= endAngle:
				rad = math.radians(a)
				x_inner = cx + innerR * math.cos(rad)
				y_inner = cy + innerR * math.sin(rad)
				x_outer = cx + outerR * math.cos(rad)
				y_outer = cy + outerR * math.sin(rad)
				mx = (x_inner + x_outer) / 2.0
				my = (y_inner + y_outer) / 2.0
				rot_angle = math.radians(45.0)
				cos_r = math.cos(rot_angle)
				sin_r = math.sin(rot_angle)
				dx_inner = x_inner - mx
				dy_inner = y_inner - my
				x_inner_rot = mx + dx_inner * cos_r - dy_inner * sin_r
				y_inner_rot = my + dx_inner * sin_r + dy_inner * cos_r
				dx_outer = x_outer - mx
				dy_outer = y_outer - my
				x_outer_rot = mx + dx_outer * cos_r - dy_outer * sin_r
				y_outer_rot = my + dx_outer * sin_r + dy_outer * cos_r
				self.drawLine(int(x_inner_rot), int(y_inner_rot),
					int(x_outer_rot), int(y_outer_rot), 1, stripeColor)
				a += stripe_spacing_deg

	def drawCircle(self, cx, cy, radius, thickness, color):
		self.drawArc(cx, cy, radius, 0.0, 360.0, thickness, color)

	def fillCircle(self, cx, cy, radius, color):
		fr = (color >> 16) & 0xFF
		fg = (color >> 8) & 0xFF
		fb = color & 0xFF

		margin = 1
		xmin = int(cx - radius - margin)
		xmax = int(cx + radius + margin)
		ymin = int(cy - radius - margin)
		ymax = int(cy + radius + margin)
		if xmin < 0: xmin = 0
		if ymin < 0: ymin = 0
		if xmax >= self.WIDTH:  xmax = self.WIDTH - 1
		if ymax >= self.HEIGHT: ymax = self.HEIGHT - 1

		for y in range(ymin, ymax + 1):
			for x in range(xmin, xmax + 1):
				px = x + 0.5 - (cx + 0.5)
				py = y + 0.5 - (cy + 0.5)
				dist = math.sqrt(px * px + py * py)
				r = dist - radius
				if r <= -1.0:
					alpha = 1.0
				elif r >= 1.0:
					alpha = 0.0
				else:
					alpha = 1.0 - (r + 1.0) * 0.5
				if alpha <= 0.0:
					continue
				idx = (y * self.WIDTH + x) * 4
				self._blend(idx, fr, fg, fb, alpha)

	def drawCurvedChar(self, cx, cy, radius, angle, char, font_filename, fontsize, rotation=0):
		scale = 4
		big_size = int(fontsize * scale)
		cache_key = (font_filename, big_size)
		font = self._font_cache.get(cache_key)
		if font is None:
			font = ImageFont.truetype(font_filename, big_size)
			self._font_cache[cache_key] = font
		img = Image.new('RGBA', (big_size * 2, big_size * 2), (0, 0, 0, 0))
		draw = ImageDraw.Draw(img)
		bbox = draw.textbbox((0, 0), char, font=font)
		tw = bbox[2] - bbox[0]
		th = bbox[3] - bbox[1]
		tx = (img.width - tw) // 2 - bbox[0]
		ty = (img.height - th) // 2 - bbox[1]
		draw.text((tx, ty), char, fill=(0, 0, 0, 255), font=font)
		rot_angle = -angle - 90.0 + rotation
		rotated = img.rotate(rot_angle, expand=False, resample=Image.BICUBIC)
		final_size = int(fontsize * 1.2)
		resized = rotated.resize((final_size, final_size), Image.LANCZOS)
		x = int(cx + radius * math.cos(math.radians(angle)) - final_size // 2)
		y = int(cy + radius * math.sin(math.radians(angle)) - final_size // 2)
		for iy in range(resized.height):
			for ix in range(resized.width):
				r, g, b, a = resized.getpixel((ix, iy))
				if a == 0:
					continue
				px = x + ix
				py = y + iy
				if 0 <= px < self.WIDTH and 0 <= py < self.HEIGHT:
					idx = (py * self.WIDTH + px) * 4
					self._blend(idx, r, g, b, a / 255.0)

	def save(self, filename):
		from PIL import Image
		img = Image.frombytes('RGBA', (self.WIDTH, self.HEIGHT), bytes(self.buffer))
		img.save(filename)

File: java/Clock/Display.java
package Clock;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import java.util.HashMap;
import java.util.Map;

public class Display {
	private static final Map<String, Paint> paintCache = new HashMap<>();

	public final int WIDTH;
	public final int HEIGHT;
	public final byte[] buffer;

	public Display() {
		this(240);
	}

	public Display(int size) {
		this.WIDTH = size;
		this.HEIGHT = size;
		this.buffer = new byte[WIDTH * HEIGHT * 4];
	}

	public void clear() {
		java.util.Arrays.fill(buffer, (byte) 0);
	}

	public void fill(int color) {
		byte r = (byte) ((color >> 16) & 0xFF);
		byte g = (byte) ((color >> 8) & 0xFF);
		byte b = (byte) (color & 0xFF);
		for (int i = 0; i < buffer.length; i += 4) {
			buffer[i] = r;
			buffer[i + 1] = g;
			buffer[i + 2] = b;
			buffer[i + 3] = (byte) 255;
		}
	}

	// ARGB pixels, ready for Bitmap.setPixels
	public int[] toArgb() {
		int[] pixels = new int[WIDTH * HEIGHT];
		for (int i = 0; i < pixels.length; i++) {
			int idx = i * 4;
			pixels[i] = ((buffer[idx + 3] & 0xFF) << 24)
					| ((buffer[idx] & 0xFF) << 16)
					| ((buffer[idx + 1] & 0xFF) << 8)
					| (buffer[idx + 2] & 0xFF);
		}
		return pixels;
	}

	private void blend(int idx, int fr, int fg, int fb, double alpha) {
		if (alpha <= 0.0) {
			return;
		}
		if (alpha > 1.0) {
			alpha = 1.0;
		}

		double ba = (buffer[idx + 3] & 0xFF) / 255.0;
		double outA = alpha + ba * (1.0 - alpha);

		if (outA <= 0.0) {
			buffer[idx] = 0;
			buffer[idx + 1] = 0;
			buffer[idx + 2] = 0;
			buffer[idx + 3] = 0;
			return;
		}

		if (alpha >= 1.0) {
			buffer[idx] = (byte) fr;
			buffer[idx + 1] = (byte) fg;
			buffer[idx + 2] = (byte) fb;
			buffer[idx + 3] = (byte) 255;
			return;
		}

		int br = buffer[idx] & 0xFF;
		int bg = buffer[idx + 1] & 0xFF;
		int bb = buffer[idx + 2] & 0xFF;
		double remainder = ba * (1.0 - alpha);

		double outR = (fr * alpha + br * remainder) / outA;
		double outG = (fg * alpha + bg * remainder) / outA;
		double outB = (fb * alpha + bb * remainder) / outA;

		buffer[idx] = (byte) (int) (outR + 0.5);
		buffer[idx + 1] = (byte) (int) (outG + 0.5);
		buffer[idx + 2] = (byte) (int) (outB + 0.5);
		buffer[idx + 3] = (byte) (int) (outA * 255.0 + 0.5);
	}

	private static double mod(double x, double m) {
		double r = x % m;
		return r < 0 ? r + m : r;
	}

	// r is the signed distance from the shape edge, in pixels
	private static double coverage(double r) {
		if (r <= -1.0) return 1.0;
		if (r >= 1.0) return 0.0;
		return 1.0 - (r + 1.0) * 0.5;
	}

	private static double pointToSegmentDistance(double px, double py, double x1, double y1, double x2, double y2) {
		double dx = x2 - x1;
		double dy = y2 - y1;
		double lenSq = dx * dx + dy * dy;
		if (lenSq == 0.0) {
			double ex = px - x1;
			double ey = py - y1;
			return Math.sqrt(ex * ex + ey * ey);
		}
		double t = ((px - x1) * dx + (py - y1) * dy) / lenSq;
		if (t < 0.0) t = 0.0;
		if (t > 1.0) t = 1.0;
		double projx = x1 + t * dx;
		double projy = y1 + t * dy;
		double ex = px - projx;
		double ey = py - projy;
		return Math.sqrt(ex * ex + ey * ey);
	}

	private static double pointToArcDistance(double px, double py, double cx, double cy, double radius,
			double startAngle, double endAngle, double halfWidth) {
		double dx = px - cx;
		double dy = py - cy;
		double distToCenter = Math.sqrt(dx * dx + dy * dy);

		if (endAngle - startAngle >= 360.0) {
			return Math.abs(distToCenter - radius);
		}

		double sr = mod(Math.toRadians(startAngle), 2.0 * Math.PI);
		double er = mod(Math.toRadians(endAngle), 2.0 * Math.PI);

		boolean inRange;
		if (distToCenter == 0.0) {
			double a0 = 0.0;
			if (startAngle <= endAngle) {
				inRange = a0 >= startAngle && a0 <= endAngle;
			} else {
				inRange = a0 >= startAngle || a0 <= endAngle;
			}
		} else {
			double angle = Math.atan2(dy, dx);
			if (angle < 0.0) {
				angle += 2.0 * Math.PI;
			}
			if (sr <= er) {
				inRange = angle >= sr && angle <= er;
			} else {
				inRange = angle >= sr || angle <= er;
			}
		}

		if (inRange) {
			return Math.abs(distToCenter - radius);
		}

		double innerR = radius - halfWidth;
		double outerR = radius + halfWidth;

		double sxi = cx + innerR * Math.cos(sr);
		double syi = cy + innerR * Math.sin(sr);
		double sxo = cx + outerR * Math.cos(sr);
		double syo = cy + outerR * Math.sin(sr);

		double exi = cx + innerR * Math.cos(er);
		double eyi = cy + innerR * Math.sin(er);
		double exo = cx + outerR * Math.cos(er);
		double eyo = cy + outerR * Math.sin(er);

		double d1 = pointToSegmentDistance(px, py, sxi, syi, sxo, syo);
		double d2 = pointToSegmentDistance(px, py, exi, eyi, exo, eyo);
		double dmin = d1 < d2 ? d1 : d2;

		return dmin + halfWidth;
	}

	public void drawLine(int x1, int y1, int x2, int y2, double thickness, int color) {
		int fr = (color >> 16) & 0xFF;
		int fg = (color >> 8) & 0xFF;
		int fb = color & 0xFF;

		double hw = thickness * 0.5;
		int xmin = Math.min(x1, x2) - (int) (hw + 1.0);
		int xmax = Math.max(x1, x2) + (int) (hw + 1.0);
		int ymin = Math.min(y1, y2) - (int) (hw + 1.0);
		int ymax = Math.max(y1, y2) + (int) (hw + 1.0);

		if (xmin < 0) xmin = 0;
		if (ymin < 0) ymin = 0;
		if (xmax >= WIDTH) xmax = WIDTH - 1;
		if (ymax >= HEIGHT) ymax = HEIGHT - 1;

		for (int y = ymin; y <= ymax; y++) {
			for (int x = xmin; x <= xmax; x++) {
				double px = x + 0.5;
				double py = y + 0.5;
				double dist = pointToSegmentDistance(px, py, x1 + 0.5, y1 + 0.5, x2 + 0.5, y2 + 0.5);
				double alpha = coverage(dist - hw);
				if (alpha <= 0.0) {
					continue;
				}
				blend((y * WIDTH + x) * 4, fr, fg, fb, alpha);
			}
		}
	}

	public void drawArc(int cx, int cy, double radius, double startAngle, double endAngle, double thickness, int color) {
		drawArc(cx, cy, radius, startAngle, endAngle, thickness, color, false, 0xffffff);
	}

	public void drawArc(int cx, int cy, double radius, double startAngle, double endAngle, double thickness,
			int color, boolean stripe) {
		drawArc(cx, cy, radius, startAngle, endAngle, thickness, color, stripe, 0xffffff);
	}

	public void drawArc(int cx, int cy, double radius, double startAngle, double endAngle, double thickness,
			int color, boolean stripe, int stripeColor) {
		int fr = (color >> 16) & 0xFF;
		int fg = (color >> 8) & 0xFF;
		int fb = color & 0xFF;
		double hw = thickness * 0.5;
		int margin = (int) (hw + 1.0);
		int xmin = (int) (cx - radius - margin);
		int xmax = (int) (cx + radius + margin);
		int ymin = (int) (cy - radius - margin);
		int ymax = (int) (cy + radius + margin);
		if (xmin < 0) xmin = 0;
		if (ymin < 0) ymin = 0;
		if (xmax >= WIDTH) xmax = WIDTH - 1;
		if (ymax >= HEIGHT) ymax = HEIGHT - 1;
		for (int y = ymin; y <= ymax; y++) {
			for (int x = xmin; x <= xmax; x++) {
				double px = x + 0.5;
				double py = y + 0.5;
				double dist = pointToArcDistance(px, py, cx + 0.5, cy + 0.5, radius, startAngle, endAngle, hw);
				double alpha = coverage(dist - hw);
				if (alpha <= 0.0) {
					continue;
				}
				blend((y * WIDTH + x) * 4, fr, fg, fb, alpha);
			}
		}

		if (stripe) {
			double halfWidth = thickness / 2.0;
			double innerR = radius - halfWidth;
			double outerR = radius + halfWidth;
			double stripeSpacingDeg = 6.0;
			double gridOrigin = -90.0;
			double n0 = Math.ceil((startAngle - gridOrigin) / stripeSpacingDeg);
			double a = gridOrigin + n0 * stripeSpacingDeg;
			double cosR = Math.cos(Math.toRadians(45.0));
			double sinR = Math.sin(Math.toRadians(45.0));
			while (a <= endAngle) {
				double rad = Math.toRadians(a);
				double xInner = cx + innerR * Math.cos(rad);
				double yInner = cy + innerR * Math.sin(rad);
				double xOuter = cx + outerR * Math.cos(rad);
				double yOuter = cy + outerR * Math.sin(rad);
				double mx = (xInner + xOuter) / 2.0;
				double my = (yInner + yOuter) / 2.0;
				double dxInner = xInner - mx;
				double dyInner = yInner - my;
				double xInnerRot = mx + dxInner * cosR - dyInner * sinR;
				double yInnerRot = my + dxInner * sinR + dyInner * cosR;
				double dxOuter = xOuter - mx;
				double dyOuter = yOuter - my;
				double xOuterRot = mx + dxOuter * cosR - dyOuter * sinR;
				double yOuterRot = my + dxOuter * sinR + dyOuter * cosR;
				drawLine((int) xInnerRot, (int) yInnerRot, (int) xOuterRot, (int) yOuterRot, 1, stripeColor);
				a += stripeSpacingDeg;
			}
		}
	}

	public void drawCircle(int cx, int cy, double radius, double thickness, int color) {
		drawArc(cx, cy, radius, 0.0, 360.0, thickness, color);
	}

	public void fillCircle(int cx, int cy, double radius, int color) {
		int fr = (color >> 16) & 0xFF;
		int fg = (color >> 8) & 0xFF;
		int fb = color & 0xFF;

		int margin = 1;
		int xmin = (int) (cx - radius - margin);
		int xmax = (int) (cx + radius + margin);
		int ymin = (int) (cy - radius - margin);
		int ymax = (int) (cy + radius + margin);
		if (xmin < 0) xmin = 0;
		if (ymin < 0) ymin = 0;
		if (xmax >= WIDTH) xmax = WIDTH - 1;
		if (ymax >= HEIGHT) ymax = HEIGHT - 1;

		for (int y = ymin; y <= ymax; y++) {
			for (int x = xmin; x <= xmax; x++) {
				double px = x + 0.5 - (cx + 0.5);
				double py = y + 0.5 - (cy + 0.5);
				double dist = Math.sqrt(px * px + py * py);
				double alpha = coverage(dist - radius);
				if (alpha <= 0.0) {
					continue;
				}
				blend((y * WIDTH + x) * 4, fr, fg, fb, alpha);
			}
		}
	}

	// Draws one glyph centred on a point at (radius, angle) around (cx, cy), rotated by rotation degrees
	public void drawCurvedChar(int cx, int cy, double radius, double angle, String ch, Typeface typeface,
			int fontSize, double rotation) {
		drawCurvedChar(cx, cy, radius, angle, ch, typeface, fontSize, rotation, Color.BLACK);
	}

	// color is ARGB
	public void drawCurvedChar(int cx, int cy, double radius, double angle, String ch, Typeface typeface,
			int fontSize, double rotation, int color) {
		int scale = 4;
		int bigSize = (int) (fontSize * scale);
		String cacheKey = System.identityHashCode(typeface) + ":" + bigSize + ":" + color;
		Paint paint = paintCache.get(cacheKey);
		if (paint == null) {
			paint = new Paint(Paint.ANTI_ALIAS_FLAG);
			paint.setTypeface(typeface);
			paint.setTextSize(bigSize);
			paint.setColor(color);
			paintCache.put(cacheKey, paint);
		}

		int side = bigSize * 2;
		Bitmap img = Bitmap.createBitmap(side, side, Bitmap.Config.ARGB_8888);
		Canvas canvas = new Canvas(img);
		Rect bbox = new Rect();
		paint.getTextBounds(ch, 0, ch.length(), bbox);
		float tx = (side - bbox.width()) / 2 - bbox.left;
		float ty = (side - bbox.height()) / 2 - bbox.top;

		// Canvas rotates clockwise, the Python (PIL) rotation was counter-clockwise
		double rotAngle = -angle - 90.0 + rotation;
		canvas.save();
		canvas.rotate((float) -rotAngle, side / 2f, side / 2f);
		canvas.drawText(ch, tx, ty, paint);
		canvas.restore();

		int finalSize = (int) (fontSize * 1.2);
		Bitmap resized = Bitmap.createScaledBitmap(img, finalSize, finalSize, true);
		int[] pixels = new int[finalSize * finalSize];
		resized.getPixels(pixels, 0, finalSize, 0, 0, finalSize, finalSize);
		if (resized != img) {
			resized.recycle();
		}
		img.recycle();

		int x = (int) (cx + radius * Math.cos(Math.toRadians(angle)) - finalSize / 2);
		int y = (int) (cy + radius * Math.sin(Math.toRadians(angle)) - finalSize / 2);
		for (int iy = 0; iy < finalSize; iy++) {
			for (int ix = 0; ix < finalSize; ix++) {
				int argb = pixels[iy * finalSize + ix];
				int a = (argb >>> 24) & 0xFF;
				if (a == 0) {
					continue;
				}
				int px = x + ix;
				int py = y + iy;
				if (px >= 0 && px < WIDTH && py >= 0 && py < HEIGHT) {
					blend((py * WIDTH + px) * 4, (argb >> 16) & 0xFF, (argb >> 8) & 0xFF, argb & 0xFF, a / 255.0);
				}
			}
		}
	}
}
